ocdskit/oc4ids_transforms.py

```diff
@@ -0,0 +1,1010 @@
+import copy
+import datetime
+import json
+import logging
+from collections import defaultdict
+
+from jsonpointer import resolve_pointer
+from ocdsmerge.util import sorted_releases
+
+from ocdskit.combine import merge
+
+logger = logging.getLogger("ocdskit")
+
+
+def check_type(item, item_type):
+    """ Check type and if incorrect return empty version of type so that future processing works with bad data"""
+    if not isinstance(item, item_type):
+        if item:
+            logger.warn("item {} is not of type {} so skipping".format(item, item_type.__name__))
+        return item_type()
+    return item
+
+
+def cast(item, item_type):
+    """ Cast to type and if casting fails return empty version of type """
+    try:
+        return item_type(item)
+    except ValueError:
+        return item_type()
+
+
+def run_transforms(config, releases, project_id=None, records=None, output=None):
+
+    """
+    Transforms a list of OCDS releases into a OC4IDS project.
+
+    :param dict config: contains optional tranform options.
+    :param list releases: list of OCDS releases
+    :param string project_id: project ID of resulting project
+    :param list records: pre computed list of records
+    :param dict output: initial project output template project where transformed data will be added
+    """
+    transforms_to_run = []
+
+    for transform in TRANSFORM_LIST:
+        transform_name = transform.__name__
+        if not config.get("all") and not config.get(transform_name) and transform_name in OPTIONAL_TRANSFORMS:
+            continue
+        transforms_to_run.append(transform)
+
+    return _run_transforms(releases, project_id, records, output, transforms_to_run)
+
+
+def _run_transforms(releases, project_id=None, records=None, output=None, transforms=None):
+
+    state = InitialTransformState(releases, project_id, records, output)
+    if not transforms:
+        transforms = TRANSFORM_LIST
+
+    for transform in transforms:
+        transform(state)
+
+    return state.output
+
+
+class InitialTransformState:
+    def __init__(self, releases, project_id=None, records=None, output=None):
+        self.releases = sorted_releases(releases)
+        self.releases_by_ocid = defaultdict(list)
+        for release in self.releases:
+            ocid = check_type(release.get("ocid"), str)
```

Comment:
Why not `cast(…, str)` (same for all cases of `check_type(…, str)`)? If they put an integer, then it's fine to just coerce it to a string. The only issue is if the value is `None`, in which case we don't want it coerced to `"None"`. We could have a new method that accepts ints, floats, strs and coerces them to str, and otherwise behaves like `check_type`.

I think it's worthwhile to add a heredoc to explain that `check_type` only makes sense for list, dict, str (as otherwise it creates a 0-value int or float, which is probably not desired).

Similarly, `cast` probably only makes sense for int and float, as otherwise you can make a str out of anything, which is again probably not what you want.

Reply:
I have changed the casting/checking functions more in line with your suggestion.

For the int/float one we did actually want it to return zero as they are used in summing.  Now this function is more explicit.  

Could you check if this is what you were expecting.